presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DereferencePushDownProject.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.presto.metadata.Metadata;
+import com.facebook.presto.sql.parser.SqlParser;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.iterative.Rule.Context;
+import com.facebook.presto.sql.planner.plan.Assignments;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.ProjectNode;
+import com.facebook.presto.sql.tree.DereferenceExpression;
+import com.facebook.presto.sql.tree.Expression;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.List;
+import java.util.Map;
+
+import static com.facebook.presto.sql.planner.SymbolsExtractor.extractAll;
+import static com.facebook.presto.sql.planner.plan.Patterns.project;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static com.google.common.collect.Iterables.getOnlyElement;
+
+public class DereferencePushDownProject
+        extends DereferencePushDownRule<ProjectNode>
+{
+    public DereferencePushDownProject(Metadata metadata, SqlParser sqlParser)
+    {
+        super(metadata, sqlParser, project());
+    }
+
+    @Override
+    protected PlanNode dereferencePushDown(Context context, ProjectNode projectNode, Map<DereferenceExpression, Symbol> expressions, Assignments assignments)
+    {
+        List<Symbol> outputSymbols = projectNode.getOutputSymbols();
+        Map<Expression, Symbol> pushdownExpressions = expressions.entrySet().stream()
+                .filter(entry -> !isCastRowType(entry.getKey()))
+                .filter(entry -> outputSymbols.contains(getOnlyElement(extractAll(entry.getKey()))))
+                .collect(toImmutableMap(Map.Entry::getKey, Map.Entry::getValue));
+
+        if (pushdownExpressions.isEmpty()) {
+            return projectNode;
+        }
+
+        Map<Symbol, Expression> pushdownDereferences = pushdownExpressions.entrySet().stream().collect(toImmutableMap(Map.Entry::getValue, Map.Entry::getKey));
```

Comment:
This variable is not needed. You can swap key and value in pushdownExpressions instead.

```
        Map<Symbol, DereferenceExpression> pushdownExpressions = expressions.entrySet().stream()
                .filter(entry -> !isCastRowType(entry.getKey()))
                .filter(entry -> outputSymbols.contains(getOnlyElement(extractAll(entry.getKey()))))
                .collect(toImmutableMap(Map.Entry::getValue, Map.Entry::getKey));
...
        ImmutableMap.Builder<Symbol, Symbol> symbolsBuilder = ImmutableMap.builder();
        pushdownExpressions.entrySet().stream()
                .forEach(entry -> symbolsBuilder.put(getOnlyElement(extractAll(entry.getValue())), entry.getKey()));
        Map<Symbol, Symbol> symbolsMap = symbolsBuilder.build();
...
        DereferenceExpression targetDereference = pushdownExpressions.get(targetSymbol);
```